plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/DateCoercer.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.coercions;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.type.DateType;
+import io.trino.spi.type.VarcharType;
+
+import java.time.LocalDate;
+import java.time.chrono.IsoChronology;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.format.DateTimeParseException;
+import java.util.regex.Pattern;
+
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
+import static java.time.format.ResolverStyle.STRICT;
+
+public final class DateCoercer
+{
+    private static final Pattern ISO_FORMAT_DATE_PATTERN = Pattern.compile("^\\d{4,10}-\\d{2}-\\d{2}$");
+    private static final DateTimeFormatter DATE_FORMATTER = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .append(ISO_LOCAL_DATE)
+            .toFormatter()
+            .withResolverStyle(STRICT)
+            .withChronology(IsoChronology.INSTANCE);
+
+    private DateCoercer() {}
+
+    public static class VarcharToDateCoercer
+            extends TypeCoercer<VarcharType, DateType>
+    {
+        public VarcharToDateCoercer(VarcharType fromType, DateType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            boolean validFormat = false;
+            String value = fromType.getSlice(block, position).toStringUtf8();
+            try {
+                validFormat = ISO_FORMAT_DATE_PATTERN.matcher(value).matches();
+                LocalDate localDate = DATE_FORMATTER.parse(value, LocalDate::from);
+                toType.writeLong(blockBuilder, localDate.toEpochDay());
+            }
+            catch (DateTimeParseException ignored) {
+                if (validFormat) {
```

Comment:
Why should we throw an exception when the date is invalid?
I see in case of the other coercers that they simply fallback to `null` instead.

Reply:
The idea is to support the same behaviour as hive or throw.
I will add more details to the pr description then

Reply:
```
trino> select * from hive.default.testp1;
 c1 |     c2     | id 
----+------------+----
  3 | 2023-03-03 | 30 
(1 row)

Query 20231003_222033_00003_6f4hh, FAILED, 1 node
http://localhost:8080/ui/query.html?20231003_222033_00003_6f4hh
Splits: 3 total, 2 done (66.67%)
CPU Time: 0.5s total,     3 rows/s, 1.64KB/s, 4% active
Per Node: 0.0 parallelism,     0 rows/s,    35B/s
Parallelism: 0.0
Peak Memory: 192B
24.62 [2 rows, 856B] [0 rows/s, 35B/s]

Query 20231003_222033_00003_6f4hh failed: Invalid date value: 2023-33-33 is not a valid date
java.lang.IllegalArgumentException: Invalid date value: 2023-33-33 is not a valid date
	at io.trino.plugin.hive.coercions.DateCoercer$VarcharToDateCoercer.applyCoercedValue(DateCoercer.java:64)
	at io.trino.plugin.hive.coercions.TypeCoercer.apply(TypeCoercer.java:46)

```

This doesn't look good from UX perspective - notice that part of the answer has been already delivered before the exception is being thrown.

Reply:
ok I will change implementation to return null if it is impossible to get a value that makes sense

Reply:
we must not return NULL when Hive returns something else.

Reply:
if Trino query completes, it must have produced same results as Hive.

we either replicate all the lenient conversions OR (preferred) strictly validate data to avoid amiguity.
Lenient conversions can even be an opt-in.